#! /bin/sh /usr/share/dpatch/dpatch-run
## fix-insufficient-checking-leading-to-privilege-escalation-str4450.dpatch by Michael Sweet <msweet@apple.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Description: The web interface incorrectly served symlinked files and files
## DP:  that were not world-readable, potentially leading to a disclosure of
## DP:  information
## DP: Author: Michael Sweet <msweet@apple.com>
## DP: Bug-CVE: https://security-tracker.debian.org/tracker/CVE-2014-3537
## DP: Bug: http://www.cups.org/str.php?L4450
## DP: Last-Update: 2014-07-14

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' cups~/scheduler/client.c cups/scheduler/client.c
--- cups~/scheduler/client.c	2014-07-30 08:33:36.000000000 +0200
+++ cups/scheduler/client.c	2014-07-30 09:00:07.000000000 +0200
@@ -3568,7 +3568,7 @@
     if ((ptr = strchr(filename, '?')) != NULL)
       *ptr = '\0';
 
-    if ((status = stat(filename, filestats)) != 0)
+    if ((status = lstat(filename, filestats)) != 0)
     {
      /*
       * Drop the language prefix and try the root directory...
@@ -3580,12 +3580,33 @@
       if ((ptr = strchr(filename, '?')) != NULL)
 	*ptr = '\0';
 
-      status = stat(filename, filestats);
+      status = lstat(filename, filestats);
     }
   }
 
  /*
-  * If we're found a directory, get the index.html file instead...
+  * If we've found a symlink, 404 the sucker to avoid disclosing information.
+  */
+
+  if (!status && S_ISLNK(filestats->st_mode))
+  {
+    cupsdLogMessage(CUPSD_LOG_INFO, "[Client %d] Symlinks such as \"%s\" are not allowed.", con->http.fd, filename);
+    return (NULL);
+  }
+
+ /*
+  * Similarly, if the file/directory does not have world read permissions, do
+  * not allow access...
+  */
+
+  if (!status && !(filestats->st_mode & S_IROTH))
+  {
+    cupsdLogMessage(CUPSD_LOG_INFO, "[Client %d] Files/directories such as \"%s\" must be world-readable.", con->http.fd, filename);
+    return (NULL);
+  }
+
+ /*
+  * If we've found a directory, get the index.html file instead...
   */
 
   if (!status && S_ISDIR(filestats->st_mode))
